core/src/main/java/hudson/model/Queue.java

```diff
@@ -1372,6 +1375,18 @@ public void maintain() {
                     makePending(p);
                 else
                     LOGGER.log(Level.FINE, "BuildableItem {0} with empty work units!?", p);
+                                
+                // Ensure that identification of blocked tasks is using the live state: JENKINS-27708 & JENKINS-27871
+                // The creation of a snapshot itself should be relatively cheap given the expected rate of
+                // job execution. You probably would need 100's of jobs starting execution every iteration
+                // of maintain() before this could even start to become an issue and likely the calculation
+                // of isBuildBlocked(p) will become a bottleneck before updateSnapshot() will. Additionally
+                // since the snapshot itself only ever has at most one reference originating outside of the stack
+                // it should remain in the eden space and thus be cheap to GC.
+                // See https://issues.jenkins-ci.org/browse/JENKINS-27708?focusedCommentId=225819&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-225819
+                // or https://issues.jenkins-ci.org/browse/JENKINS-27708?focusedCommentId=225906&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-225906
+                // for alternative fixes of this issue.
+                updateSnapshot();
```

Comment:
The finally block could probably be split, so that here is the end of the inner-most finally but it would have applied a whole host of formatting changes... and given I want to have this as `lts-candidate` in the event that 1.607-1.609 are selected for the next LTS I though this was least risk